Refactor Policy and Rule resources

quantumew opened the pull request:
The original issue here was to detect and correct drift, which then made us realize we were not calling Read from Update and Create per Terraform best practices, which uncovered a load of attributes that were not being set right but the tests passed due to stale state. While correcting this it became immediately obvious that these had over complicated schemas and thus the refactor.

Reasons for splitting up Policy types?
- No longer had to deal with "conditional defaults" or "conditional requirements
- No longer have inconsistent schema due to different policy types requiring or disallowing certain attributes
- Still have centralized functionality

Reasons for refactor in general?
- A lot of copy code that I was able to centralize
- Needlessly nested attributes
- Lists of Maps that should have just been Maps
- Easier to consume

Once code review is complete and before merge I will update tf-rise-okta

Fixes #70 Fixes #61

tecnobrat commented on okta/provider_sweeper_test.go:
```diff
@@ -0,0 +1,75 @@
+package okta
```
Whats this sweeper thing? My understanding is that you can write your tests in a way that this doesn't happen?

Your tests should destroy after, and actually test to ensure it happened? https://github.com/articulate/terraform-provider-kong/blob/master/kong/resource_kong_api_test.go#L24

tecnobrat replied:
![image](https://user-images.githubusercontent.com/145406/45779803-e575c780-bc10-11e8-92e4-fc76b620dab4.png)

quantumew replied:
https://www.terraform.io/docs/extend/testing/acceptance-tests/sweepers.html

The sweepers are supposed to be a failsafe for leaks. They were easy to implement and I saw a lot of providers implementing them so I added it as well. And yea totally, the old destroy logic also needs work, haven't gotten to refactoring this yet. It works fine with the happy path but I was seeing leaks in certain cases. I will fix that!

tecnobrat replied:
Ahh I see, so this is to handle the case where a previous test may have failed and not cleaned up.

quantumew replied:
Yea sweepers are a failsafe to avoid dangling resources, which I was seeing a lot of when tests would fail. I will take a bit of time on this PR and try fixing the destroy logic so it works on failure as well, if that is possible. I bet it is.

tecnobrat commented:
Looks great!

I just had the one question about the sweeper stuff, seems like that goes against the patterns TF has about cleaning up after itself.

tecnobrat commented on GNUmakefile:
```diff
@@ -1,5 +1,6 @@
 VERSION=v1.1.5
 
+SWEEP?=us-east-1
```
`us-east-1`?

was this stolen from an AWS provider or something? :)

quantumew replied:
Yea just a dumby value to enable sweeping. They had no regionless implementation of the sweepers so I just hardcode one and ignore it in the sweep functions.

quantumew commented on okta/policy.go:
```diff
@@ -0,0 +1,150 @@
+		Optional:    true,
+		Description: "Policy Description",
+	},
+	// Need to add an enhancement here, maybe we can provide a hierarchy system, priorityBefore = <refPolicy>
```
Chatted with Steve, he had a good point on this, while it is hard to make this complete dynamic, we can have a hard failure when priority cannot be set (usually due to priority being set to a value higher then the number of resources). Implement that and remove this comment.

quantumew commented:
Thanks for the feedback @tecnobrat! Updated the dumby value to "global" (API requires some value). I also found the cause of the dangling resources, there were tests that created a resource, then update it to elicit an error but didn't force a Destroy so I fixed that. I will keep sweepers around for API errors or panics after resource creation, etc.

@PurrBiscuit Updated the schema to allow you to set a priority but fail if priority is computed by API to avoid the perpetual diff situation.
